// models/Member/transport.ts
import { ITransport } from 'transport';
import { IAwaitable, awaitAll } from 'await';
import { Member, Event } from 'models';
import { Options } from './list';
import list from './list';

export interface MemberData {
  id: number;
  name: string;
  image: string;
}

interface ListOptions extends Options {
  event?: Event;
}

export default class implements ITransport<Member>, IAwaitable {

  constructor(private readonly mapResult: (data: MemberData) => Member) {}

  private readonly transport: ITransport<MemberData> = { list };
  private readonly awaiting = new Set<Promise<any>>();

  async list(opts: ListOptions = {}) {
    if (opts.event)
      opts = Object.assign({ eventId: opts.event.id }, opts);
    const promise = this.transport.list(opts);
    this.awaiting.add(promise);
    const response = await promise;
    this.awaiting.delete(promise);
    return response.map(this.mapResult);
  }

  get await() {
    return awaitAll([...this.awaiting]);
  }

  toJSON() {
    return;
  }

}
