// models/Member/transport.ts
import Transport from 'Transport';
import { Awaitable, awaitAll } from 'Awaitable';
import { Member, Event } from 'models';
import { ListOptions, list } from './list';

export interface MemberData {
  id: number;
  name: string;
  image: string;
}

interface MemberListOptions extends ListOptions {
  event?: Event;
}

export class MemberTransport implements Transport<Member>, Awaitable {

  constructor(private readonly mapResult: (data: MemberData) => Member) {}

  private readonly transport: Transport<MemberData> = { list };
  private readonly awaiting = new Set<Promise<any>>();

  async list(opts: MemberListOptions = {}) {
    if (opts.event)
      opts = Object.assign({ eventId: opts.event.id }, opts);
    const promise = this.transport.list(opts);
    this.awaiting.add(promise);
    const response = await promise;
    this.awaiting.delete(promise);
    return response.map(this.mapResult);
  }

  get await() {
    return awaitAll([...this.awaiting]);
  }

  toJSON() {
    return;
  }

}

export default MemberTransport;
